### PR TITLE
Address using ArraySequence objects instead of arrays

### PR DESCRIPTION
**Pull request by giorgiosironi**
I only updated the tests of Address itself, to get some feedback. Is this the right direction?

### CODE REVIEW

File: src/Serializer/AddressNormalizer.php [hunk 1]
```diff
@@ -28,20 +31,20 @@ public function supportsDenormalization($data, $type, $format = null) : bool
...
     {
         $data = [
-            'formatted' => $object->getFormatted(),
+            'formatted' => $object->getFormatted()->toArray(),
             'components' => [],
         ];
 
-        if ($object->getStreetAddress()) {
-            $data['components']['streetAddress'] = $object->getStreetAddress();
+        if (!$object->getStreetAddress()->isEmpty()) {
```
Comment (giorgiosironi) [inline src/Serializer/AddressNormalizer.php]: maybe we need some `isThere()` method that is the opposite of `isEmpty()`, to be placed on `CanBeCounted`

Comment (thewilkybarkid) [reply src/Serializer/AddressNormalizer.php]: If it were `if (!$streetAddress->isEmpty())` I'd say it's ok, but the `!` is a bit disconnected here. Maybe `notEmpty()`?

File: src/Model/Address.php [hunk 2]
```diff
@@ -15,49 +18,49 @@
...
+        $this->streetAddress = ArraySequence::coalesce($streetAddress);
+        $this->locality = ArraySequence::coalesce($locality);
+        $this->area = ArraySequence::coalesce($area);
         $this->country = $country;
         $this->postalCode = $postalCode;
     }
 
     /**
-     * @return string[]
+     * @return Sequence of string
```
Comment (thewilkybarkid) [inline src/Model/Address.php]: `@return Sequence|string[]`

File: src/Collection/ArraySequence.php [hunk 3]
```diff
@@ -16,6 +16,15 @@
 
     private $array;
 
+    public static function coalesce(self $value = null) : self
```
Comment (thewilkybarkid) [inline src/Collection/ArraySequence.php]: Maybe we should have `sequence_for()` function (similar to Guzzle's `promise_for()`) that returns the appropriate type (`PromiseSequence` for a `PromiseInterface`, `ArraySequence` for an `array` or any single value).

File: src/Model/Address.php [hunk 4]
```diff
@@ -15,49 +18,49 @@
...
      */
     public function __construct(
-        array $formatted,
-        array $streetAddress = [],
-        array $locality = [],
-        array $area = [],
+        Sequence $formatted,
+        Sequence $streetAddress = null,
+        Sequence $locality = null,
+        Sequence $area = null,
```
Comment (thewilkybarkid) [inline src/Model/Address.php]: I've just been requiring `Sequences` elsewhere, and passing in empty values.

File: src/Model/Address.php (hunk 2 above)
Comment (giorgiosironi) [reply src/Model/Address.php]: doesn't that mean "returns a `Sequence` or a `string` array"?

Comment (thewilkybarkid) [reply src/Model/Address.php]: It does, but the return type is forced to `Sequence`. I put it like that to help IDEs with typehinting (eg for doing a `foreach()`). If only PHPDoc was better.

**Comment by giorgiosironi** [comment]
Now ready for review, before applying similar changes to other Models.

**Pull request merged**